### PR TITLE
Port master_update_shard_statistics to new connection API

### PR DESCRIPTION
**Pull request by lithp**
Linked to #1101

### CODE REVIEW

File: src/backend/distributed/master/master_stage_protocol.c [hunk 1]
```diff
@@ -685,39 +683,50 @@ WorkerShardStats(char *nodeName, uint32 nodePort, Oid relationId, char *shardNam
...
+		ReportConnectionError(connection, WARNING);
+		ereport(WARNING, (errmsg("first query not send")));
 		return false;
 	}
 
-	tableSizeString = MultiClientGetValue(queryResult, 0, 0);
+	queryResult = GetRemoteCommandResult(connection, doRaiseInterrupts);
+	if (!IsResponseOK(queryResult))
+	{
+		ereport(WARNING, (errmsg("first response not OK")));
```
Comment (anarazel) [inline src/backend/distributed/master/master_stage_protocol.c]: Use ReportResultError instead.

File: src/backend/distributed/connection/remote_commands.c [hunk 2]
```diff
@@ -279,7 +279,7 @@ SendRemoteCommand(MultiConnection *connection, const char *command)
 
 
 /*
- * GetCommandResult is a wrapper around PQgetResult() that handles interrupts.
+ * GetRemoteCommandResult is a wrapper around PQgetResult() that handles interrupts.
```
Comment (anarazel) [inline src/backend/distributed/connection/remote_commands.c]: Oops.

**Comment by anarazel** [comment]
> I assume GetPlacementConnection always returns a connection, is that a good idea?

It does.  There's no guarantee however that the connection is in a good state (it's impossible to guarantee that anyway...).

**Comment by anarazel** [comment]
Btw, have I missed something, or do the contents of this PR and the title don't match?

**Comment by lithp** [comment]
For this one you're right, the title isn't correct yet :) I started there but thought it made sense to tackle WorkerShardStats first and then work my way out. My the time this PR is finished it'll have both.

Thanks for the feedback, I'll get rid of all the debug logging tomorrow.

**Comment by anarazel** [comment]
On 2017-01-12 11:51:28 -0800, Brian Cloutier wrote:
> For this one you're right, the title isn't correct yet :) I started there but thought it made sense to tackle WorkerShardStats first and then work my way out. My the time this PR is finished it'll have both.

I'd rather have them in separate PRs (so we can merge WorkerShardStats
quicker, it's a blocker for COPY)... If you need the changes for the
master_append_table_to_shard() PR, just include them, and then rebased
when WorkerShardStats is merged.

**Comment by lithp** [comment]
Okay, makes sense. This'll probably be ready by the end of tomorrow then.

**Comment by lithp** [comment]
@anarazel This is ready for you to look at.

**Comment by lithp** [comment]
@anarazel I think this is ready for you to look at while I figure out what to do about code coverage. It makes sense that the error paths should also be tested but I don't think we have the tooling required to test these error conditions.

File: src/backend/distributed/master/master_stage_protocol.c [hunk 3]
```diff
@@ -685,39 +684,51 @@ WorkerShardStats(char *nodeName, uint32 nodePort, Oid relationId, char *shardNam
...
+		ForgetResults(connection);
 		return false;
 	}
 
-	tableSizeString = MultiClientGetValue(queryResult, 0, 0);
+	tableSizeString = PQgetvalue(queryResult, 0, 0);
 	if (tableSizeString == NULL)
 	{
-		MultiClientDisconnect(connectionId);
+		ereport(WARNING, (errmsg("received bad result from a shard placement, trying the"
```
Comment (anarazel) [inline src/backend/distributed/master/master_stage_protocol.c]: I'm unconvinced that WorkerShardStats emitting this message makes sense - there might not be a next placement?

File: src/backend/distributed/master/master_stage_protocol.c [hunk 4]
```diff
@@ -728,28 +739,41 @@ WorkerShardStats(char *nodeName, uint32 nodePort, Oid relationId, char *shardNam
...
+
+	queryResult = GetRemoteCommandResult(connection, doRaiseInterrupts);
+	if (!IsResponseOK(queryResult))
 	{
-		MultiClientDisconnect(connectionId);
+		ReportResultError(connection, queryResult, WARNING);
+		PQclear(queryResult);
+		ForgetResults(connection);
 		return false;
 	}
```
Comment (anarazel) [inline src/backend/distributed/master/master_stage_protocol.c]: Might make more sense to add a wrapper for PQexec() that uses Send/GetRemoteCommandResult

File: src/backend/distributed/master/master_stage_protocol.c (hunk 3 above)
Comment (lithp) [inline src/backend/distributed/master/master_stage_protocol.c]: I've removed it but the idea was that without these messages, should this fail, we have no idea where it failed. I added slightly different verbiage to each of them so we would be able to take an error report and know exactly which condition was taken.

Without these messages we're silently discarding errors.

File: src/backend/distributed/master/master_stage_protocol.c (hunk 4 above)
Comment (lithp) [inline src/backend/distributed/master/master_stage_protocol.c]: That makes sense eventually. We already have `ExecuteCriticalRemoteCommand`, adding `ExecuteRemoteCommand` sounds reasonable.

 This pattern only occurs three times in the current codebase (including these two, the third is [here](https://github.com/citusdata/citus/blob/master/src/backend/distributed/transaction/transaction_recovery.c#L241)) and it's not clear to me how common it'll eventually be, maybe it's worth waiting?

**Comment by lithp** [comment]
@mtuncer I think this is ready for you.

File: src/backend/distributed/master/master_stage_protocol.c (hunk 4 above)
Comment (mtuncer) [inline src/backend/distributed/master/master_stage_protocol.c]: I think this is a good time to do that, we could refactor it if needed.

Comment (lithp) [inline src/backend/distributed/master/master_stage_protocol.c]: Okay, incoming

**Comment by mtuncer** [comment]
looks good to me 🚢

File: src/backend/distributed/master/master_stage_protocol.c [hunk 5]
```diff
@@ -636,7 +638,7 @@ UpdateShardStatistics(int64 shardId)
  * we assume have changed after new table data have been appended to the shard.
  */
 static bool
-WorkerShardStats(char *nodeName, uint32 nodePort, Oid relationId, char *shardName,
+WorkerShardStats(ShardPlacement *placement, Oid relationId, char *shardName,
```
Comment (ozgune) [inline src/backend/distributed/master/master_stage_protocol.c]: @lithp -- minor note. No need to take an action unless you strongly agree.

Now that we changed the function's signature, is ``PlacementShardStats()`` a better name for this function?

Comment (lithp) [inline src/backend/distributed/master/master_stage_protocol.c]: Yeah, `PlacementShardStats` does sound better, that's probably worth opening a PR for.

**Pull request merged**